### PR TITLE
feat: add chrom1/chrom2 arguments to add_pair

### PR DESCRIPTION
**Pull request by msto**
Addressing #55, it would be helpful to be able to create pairs where the reads map to different contigs.

I've changed the signature for `SamBuilder.add_pair()` to accept `chrom1` and `chrom2`, instead of `chrom`. When `chrom2` is not specified, the value passed to `chrom1` will be used for both reads.

As written, this will be a breaking change for most calls to `add_pair()`, which likely use `chrom`. To minimize side effects elsewhere, we could use `chrom` and `chrom2` instead, but this conflicts with the pattern used for all other arguments (`start1`/`start2`, etc). I think it'd be better to rip the bandaid once and have this fixed, but lmk if there's a cleaner way to make the update.

### CODE REVIEW

**Comment by msto** [comment]
This would be backwards compatible but is ugly imo

```py
def add_pair(
    # ...,
    chrom: str = sam.NO_REF_NAME,
    chrom1: Optional[str] = None,
    chrom2: Optional[str] = None,
    # ...,
) -> Tuple[AlignedSegment, AlignedSegment]:
    # ...

    # Use `chrom` for both reads if `chrom1` and `chrom2` are not specified.
    if chrom1 is None and chrom2 is None:
        chrom1 = chrom2 = chrom
    elif chrom1 is None or chrom2 is None:
        raise ValueError("When using chrom1 or chrom2, both must be specified.")


```

**Comment by tfenne** [comment]
I don't feel good about breaking back-compat for this - it will likely break 100s of tests across many code-bases.  @nh13 what do you think?

I can think of a couple of other options:

1. Define a new function that takes `chrom1` and `chrom2` and vector the other function to it.  I'm having a hard time coming up with a clear and short name though.
2. Leave `chrom` in add_pair (and perhaps deprecate it?) _and_ add `chrom1` and `chrom2` and have clear rules about precedence (or just fail if called with values for `chrom` and `chrom1`?)

Or just leave it at `chrom`/`chrom2` and accept the annoying difference from all the other fields.

I think I'd vote for (2) above.

**Comment by msto** [comment]
I updated so the following combinations of parameters are permitted:

```
# Backwards compatible:
- chrom, start1, start2
- chrom, start1
- chrom, start2

# New:
- chrom1, start1, chrom2, start2
- chrom1, start1
- chrom2, start2
```

**Review by nh13 — APPROVED** [review]
Nice, very thoughtful.

File: fgpyo/sam/builder.py [hunk 1]
```diff
@@ -405,15 +420,49 @@ def add_pair(
...
+        # - chrom1, start1, chrom2, start2
+        # - chrom1, start1
+        # - chrom2, start2
+
+        # Fail if chrom is specified with chrom1 or chrom2
+        if chrom != sam.NO_REF_NAME and (chrom1 is not None or chrom2 is not None):
+            raise ValueError("chrom cannot be specified together with chrom1 or chrom2.")
+
+        # Backwards compatibility
+        # Use `chrom` for both reads if neither `chrom1` nor `chrom2` are specified.
```
Comment (tfenne) [inline fgpyo/sam/builder.py:437]: I _think_ this is overly restrictive, but I'm having a hard time thinking through all the cases.  Would it not suffice to do something like:

```
if chrom is None and (chrom1 is not None or chrom2 is not None):
    raise ValueError("Cannot use chrom in combination with chrom1 or chrom2")

if chrom is None and chrom1 is None and chrom2 is None:
    chrom = sam.NO_REF_NAME

chrom1 = next(c for c in (chrom2, chrom) if c is not None)
chrom2 = next(c for c in (chrom1, chrom) if c is not None)
```

**Review by tfenne — COMMENTED** [review]
Might make sense to huddle and discuss how to handle chrom/chrom1/chrom2.

File: fgpyo/sam/builder.py [hunk 2]
```diff
@@ -405,15 +420,49 @@ def add_pair(
...
+        # - chrom, start1, start2
+        # - chrom, start1
+        # - chrom, start2
+        # Valid parameterizations for contig mapping (new):
+        # - chrom1, start1, chrom2, start2
+        # - chrom1, start1
+        # - chrom2, start2
+
+        # Fail if chrom is specified with chrom1 or chrom2
+        if chrom != sam.NO_REF_NAME and (chrom1 is not None or chrom2 is not None):
```
[the remark this answers is not in the document]
Comment (msto) [reply fgpyo/sam/builder.py]: I hadn't wanted to change the function signature but I agree

File: fgpyo/sam/builder.py (hunk 1 above)
Comment (msto) [reply fgpyo/sam/builder.py]: Yeah I find this clearer.

Just pushed, with additional checks to ensure a chromosome is always specified when providing a start position

https://github.com/fulcrumgenomics/fgpyo/pull/61/commits/8a55ab18d68aa57b09ec59edbada88e8ef30226f

**Review by tfenne — APPROVED** [review]

**Pull request merged**